rules.go

```diff
@@ -0,0 +1,218 @@
+	raw         string
+	ruleText    string
+	rawOptions  []string
+	regexString string
+	options     map[string]bool
+	optionsKeys []string
+	isComment   bool
+	isHTMLRule  bool
+	isException bool
+	document    bool
+	domain      map[string]bool
+}
+
+// Interfaces
+
+type RuleSet interface {
+	Allow(*Request) bool
+}
+
+// Allow methods
+
+func (rule *Rule) Allow(url string) bool {
+	// TODO
+	return false
+}
+
+//
+
+func ParseRule(ruleText string) (*Rule, error) {
+	rule := &Rule{}
+	rule.raw = ruleText
+	ruleText = strings.TrimSpace(ruleText)
+	rule.isComment = strings.Contains(ruleText, "!") || strings.Contains(ruleText, "[Adblock")
+	if rule.isComment {
+		rule.isHTMLRule = false
+		rule.isException = false
+	} else {
+		rule.isHTMLRule = strings.Contains(ruleText, "##") || strings.Contains(ruleText, "#@#")
+		rule.isException = strings.HasPrefix(ruleText, "@@")
+		if rule.isException {
+			ruleText = ruleText[2:]
+		}
+	}
+
+	rule.options = make(map[string]bool)
+	if !rule.isComment && strings.Contains(ruleText, "$") {
+		var option string
+		parts := strings.SplitN(ruleText, "$", 2)
+		length := len(parts)
+		if length > 0 {
+			ruleText = parts[0]
+		}
+		if length > 1 {
+			option = parts[1]
+		}
+
+		rule.rawOptions = strings.Split(option, ",")
+		for _, opt := range rule.rawOptions {
+			if strings.HasPrefix(opt, "domain=") {
+				rule.domain = parseDomainOption(opt)
+			} else {
+				rule.options[strings.TrimPrefix(opt, "~")] = !strings.HasPrefix(opt, "~")
+			}
+		}
+	} else {
+		rule.rawOptions = []string{}
+		rule.domain = make(map[string]bool)
+	}
+
+	rule.optionsKeys = rule.OptionsKeys()
+	rule.ruleText = ruleText
+
+	if rule.isComment || rule.isHTMLRule {
+		rule.regexString = ""
+	} else {
+		var err error
+		rule.regexString, err = ruleToRegexp(ruleText)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return rule, nil
+}
+
+func NewRuleSet(rules []*Rule) (RuleSet, error) {
+	return nil, nil
+}
+
+func (rule *Rule) OptionsKeys() []string {
+	opts := []string{}
+	for opt := range rule.options {
+		opts = append(opts, opt)
+	}
+	if rule.domain != nil && len(rule.domain) >= 0 {
+		opts = append(opts, "domain")
+	}
+	return opts
+}
+
+func parseDomainOption(text string) map[string]bool {
+	domains := text[len("domain="):]
+	parts := strings.Split(strings.Replace(domains, ",", "|", -1), "|")
+	opts := make(map[string]bool, len(parts))
+	for _, part := range parts {
+		opts[strings.TrimPrefix(part, "~")] = !strings.HasPrefix(part, "~")
+	}
+	return opts
+}
+
+func ruleToRegexp(text string) (string, error) {
+	// Convert AdBlock rule to a regular expression.
+	if text == "" {
+		return ".*", nil
+	}
+
+	// Check if the rule isn't already regexp
+	length := len(text)
+	if length >= 2 && text[:1] == "/" && text[length-1:] == "/" {
+		return text[1 : length-1], nil
+	}
+
+	// escape special regex characters
+	rule := escapeSpecialRegxp.ReplaceAllStringFunc(text, func(src string) string {
+		return fmt.Sprintf(`\%v`, src)
```

Comment:
Can we use https://golang.org/pkg/regexp/#QuoteMeta here instead?

Reply:
Very nice, much more clean